### ISSUE TITLE
Express Integration

### ISSUE THREAD

**Issue by keithbro** [henhal/openapi-ts-backend#4]
Hello! This is a great library, thank you for putting it together!

I was looking to integrate with an Express application. Would you like to have an Express connector as part of this repo or as something separate? I've started looking at writing code some to translate headers and query params etc. I'd be happy to add it here or have it separately as you wish.

**Comment by henhal** [comment]
Thanks, glad you like it.
I've been thinking about adding integrations to other frameworks too, PRs are most welcome. As you can see, lambda is supported directly in this library which isn't ideal. I was struggling whether to make add on libraries for the integrations, or if possibly using optional peerDependencies could work, but I haven't really tried that out. It's of course far from ideal that anyone using this library with lambda would get express dependencies, and vice versa. If you've got any opinions I'm all ears. 

When you say express connector to translate headers etc, is your ambition to do that similarly to how it works for the lambda subclass, ie, mount it on express app route and then implement operations with openapi-ts-backend requests and responses and the original express specific parameters (the express request I assume) in the custom data, or would you rather connect it so that the operation handlers are mapped to express handlers?

**Comment by keithbro** [comment]
>  It's of course far from ideal that anyone using this library with lambda would get express dependencies, and vice versa. If you've got any opinions I'm all ears.

Yes good point. Perhaps separate libraries are the way to go and it you can have:

`openapi-ts-backend-aws-lambda`
`openapi-ts-backend-express`

etc.

I haven't used `peerDependencies` much before, and it could work. I wonder if separate libraries is a cleaner solution ¯\_(ツ)_/¯.

In terms of how I saw the express plugin working, here is my sandbox example: https://github.com/keithbro/openapi-ts-backend-express-demo/blob/main/src/server.ts

It works similarly to yours I think although I have a `connect` function which maps an OpenApi to an Express RequestHandler: https://github.com/keithbro/openapi-ts-backend-express-demo/blob/main/src/openapi-ts-backend-express/index.ts#L20

I'm fine with changing it to a class based approach when an ExpressOpenApi would extend OpenApi, to keep it consistent with yours. Let me know what you think.

**Comment by henhal** [comment]
Nice. One potential problem with your `connect` solution is that you don't pass the original express data to your operations. They won't necessarily be needed, of course, but it could be useful. If you compare to my AWS Lambda integration, I pass the raw lambda `event` and `context` as data to the `handleRequest` method, so that operations could find things like the lambda runtime, the raw event etc, through `params.data.lambda.event`. I've had situations where I needed this.

Now, of course a `connect` based solution could pass `{req: Express.Request}` to `handleRequest()` as well, but one potential problem is that the `OpenApi` instance is still typed as `OpenApi<unknown>`, so that all the operations, authorizers and interceptors don't get typed `params`, knowing that `params.data` contains express properties. One way of typing the `OpenApi` instance is to create a sub-class, another one is to simply cast it, e.g. by having a `connect` style function return `OpenApi<ExpressParams>` or similar, after attaching the api with `express.use()`. But hiding away the `use()` call isn't ideal either so I'm not sure what the best connect signature would be, returning both an express handler function and a cast `OpenApi<ExpressParams>` instance.

An upside with having a separate connector is that you truly decouple the OpenApi stuff and the operation implementations from the HTTP framework used to serve it with requests. But at the same time, if you do need the context parameters - which depend on the framework - from your operations, then they aren't really decoupled anyway.

Anyway, this was the reason I chose the sub-class approach. If I do `new LambdaOpenApi()` I get an instance of `OpenApi<LambdaParams>` _and_ a way to get a lambda specific event handler function.
Using the same approach for express could I guess look something like

```
const api = new ExpressOpenApi().register({definition, operations});
app.use(api.requestHandler()); // method on ExpressOpenApi sub-class
```

This would force `operations` to be of type `OperationHandlers<ExpressParams>` (if the params aren't interesting of course one could declare the operations as `OperationHandlers<any>`), and operations could then do

```
listPets: async (req, res, params) => {
  // Express request is in params.data.express.req if we need it

  return [ ... ];
}
```

Note that LambdaOpenApi also has a generic type `<T>` which enables you to create an `OpenApi` instance where `params.data` contains lambda data and custom data such as `{lambda: {event, context}, myStuff: {foo, bar}}`.

Anyway, splitting up `openapi-ts-backend` is definitely the way to go. I'll start looking at a v2 version where you don't get AWS Lambda as part of that module. Perhaps the add-on modules could simply contain both styles of connections for maximum flexibility?

```
export function requestHandler(api: OpenApi<unknown>): Express.RequestHandler {
  return ...
}

export type ExpressSource {
  express: {
    request: Express.Request;
  }
};

export class ExpressOpenApi<T> extends OpenApi<ExpressSource & T> {
  requestHandler(...) { // could allow custom data as per https://github.com/henhal/openapi-ts-backend/blob/main/src/services/aws-lambda/lambda-openapi.ts#L87
    return requestHandler(this);
  }
}
```

**Comment by henhal** [comment]
@keithbro I finally took a bigger grip on this and decided to separate the openapi-ts-backend module into several modules. I created an NPM scope `@openapi-ts` to collect the related modules. This means you can now do:

```
$ npm install @openapi-ts/backend
``` 

... to get the basic module, and e.g. 

``` 
$ npm install @openapi-ts/aws-lambda
``` 
... to get the AWS Lambda specific implementation.

(I also extracted the basic types used by generated code to address #2)


I'm still working on updating the documentation and READMEs to deprecate the old module and link to the new module etc, and to make some final tests, but [here's](https://github.com/henhal/openapi-ts-aws-lambda/tree/main/src) the new module for AWS Lambda integration. 

I also created a new git for express integration [here](https://github.com/henhal/openapi-ts-express), and I wanted to ask if you if you're OK with me using parts of your example code above to create and publish this module (I will give you credit in the README in that case), or if you want to collaborate directly and add it yourself as a PR? I want to give credit where credit's due, of course!

**Comment by keithbro** [comment]
Hi @henhal ! Sorry for the radio silence. Yes of course please go ahead and use the example code as you please. I will try to find some time this weekend to contribute but by all means go ahead and don't wait for me :)

**Comment by henhal** [comment]
@keithbro Thanks, I made an initial implementation for express at https://www.npmjs.com/package/@openapi-ts/express, but TBH I haven't had time to really test it right now. :scream:  Feel free to try it out and give feedback and/or provide improvements whenever you have time. :+1: 

I'm closing this issue for now.

Cheers

**Comment by henhal** [comment]
Fixed in #5 and https://github.com/henhal/openapi-ts-express
